interfaces/builtin/posix_mq.go

```diff
@@ -0,0 +1,297 @@
+		if !iface.isValidPermission(perm) {
+			return fmt.Errorf("posix-mq slot %s permission \"%s\" not valid, must be one of %v", name, perm, posixMQPlugPermissions)
+		}
+	}
+
+	return nil
+}
+
+func (iface *posixMQInterface) getPermissions(attrs interfaces.Attrer, name string) ([]string, error) {
+	var perms []string
+
+	if err := attrs.Attr("permissions", &perms); err != nil {
+		/* If the permissions have not been specified, use the defaults */
+		perms = posixMQDefaultPlugPermissions
+	}
+
+	if err := iface.validatePermissionList(perms, name); err != nil {
+		return nil, err
+	}
+
+	return perms, nil
+}
+
+func (iface *posixMQInterface) getPath(attrs interfaces.Attrer, name string) (string, error) {
+	var path string
+
+	if pathAttr, isSet := attrs.Lookup("path"); isSet {
+		if pathStr, ok := pathAttr.(string); ok {
+			path = pathStr
+		} else {
+			return "", fmt.Errorf(`posix-mq slot %s "path" attribute must be a string, not %v`, name, pathAttr)
+		}
+	} else {
+		/* path defaults to name if unspecified */
+		path = name
+	}
+
+	/* Path must begin with a / */
+	if path[0] != '/' {
+		path = "/" + path
+	}
+
+	if err := iface.validatePath(name, path); err != nil {
+		return "", err
+	}
+
+	return path, nil
+
+}
+
+func (iface *posixMQInterface) validatePath(name, path string) error {
+	if !posixMQNamePattern.MatchString(path) {
+		return fmt.Errorf(`posix-mq slot %s "path" attribute must conform to the POSIX message queue name specifications (see "man mq_overview"): %v`, name, path)
+	}
+
+	if err := apparmor_sandbox.ValidateNoAppArmorRegexp(path); err != nil {
+		return fmt.Errorf(`posix-mq slot %s "path" attribute is invalid: %v"`, name, path)
+	}
+
+	if !cleanSubPath(path) {
+		return fmt.Errorf(`posix-mq slot %s "path" attribute is not a clean path: %v"`, name, path)
+	}
+
+	return nil
+}
+
+func (iface *posixMQInterface) BeforePreparePlug(plug *snap.PlugInfo) error {
+	if err := iface.checkPosixMQAppArmorSupport(); err != nil {
+		return err
+	}
+
+	/* Only ensure that the given permissions are valid, don't use them here */
+	if _, err := iface.getPermissions(plug, plug.Name); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (iface *posixMQInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	if err := iface.checkPosixMQAppArmorSupport(); err != nil {
+		return err
+	}
+
+	/* Only ensure that the given permissions are valid, don't use them here */
+	if _, err := iface.getPermissions(slot, slot.Name); err != nil {
+		return err
+	}
+
+	/* Only ensure that the given path is valid, don't use it here */
+	if _, err := iface.getPath(slot, slot.Name); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (iface *posixMQInterface) AppArmorPermanentSlot(spec *apparmor.Specification, slot *snap.SlotInfo) error {
+	if !implicitSystemPermanentSlot(slot) {
+		if path, err := iface.getPath(slot, slot.Name); err == nil {
+			spec.AddSnippet(fmt.Sprintf(`  # POSIX Message Queue management
+  mqueue (create delete read write) "%s",
```

Comment:
Instead of hard-coding these permissions here would it be better to just use all of the ones listed in `posixMQPlugPermissions`?

Reply:
Good point - I've changed that in the latest commit